Fix LanguageLocaleCodeProvider when working with inherited languages

silviokennecke opened the pull request:
### 1. Why is this change necessary?

When using a child language, multiple parts in shopware (like the twig currency filter) cause issues

### 2. What does this change do, exactly?

When passing a child language to `\Shopware\Core\System\Locale\LanguageLocaleCodeProvider::getLocaleForLanguageId`, the method now also checks the parent languages recursively. This ensures that the return type cannot be null, as it could right now - which isn't allowed according to the method declaration.

### 3. Describe each step to reproduce the issue or behaviour.

1. Add a new child language (`language.parentId`)
2. Open the storefront in that language

The storefront now replies with an error 500: `Return value of Shopware\Core\System\Locale\LanguageLocaleCodeProvider::getLocaleForLanguageId() must be of the type string, null returned`

### 5. Checklist

- [X] I have written tests and verified that they fail without my change
- [X] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfil them.

silviokennecke committed:
check parent locale code

silviokennecke committed:
add test

Commit:
rephrase changelog entry

OliverSkroblin commented on src/Core/System/Locale/LanguageLocaleCodeProvider.php:
```diff
@@ -28,7 +28,15 @@ public function getLocaleForLanguageId(string $languageId): string
+            $code = $this->getLocaleForLanguageId($parentId);
+        }
+
+        return $code;
```
please add this fallback for the `getLocalesForLanguageIds` too.

I would recommend that you can add this fallback directly in the language loader

silviokennecke replied:
Yes, I'll fix this in the language loader. I was just not sure if some other dependencies of `LanguageLoaderInterface::loadLanguages` expect getting `locale = null` in case of a child language.

silviokennecke committed:
revert locale code fix

silviokennecke committed:
resolve locale code for child languages

silviokennecke committed:
fix typo

keulinho commented on src/Core/System/Language/LanguageLoader.php:
```diff
@@ -26,6 +26,28 @@ public function loadLanguages(): array
+                $language['code'] = $languages[$currentLanguageId]['code'];
+            }
+
+            $languages[$languageId]['code'] = $language['code'];
```
I'm not sure what php does here if you manipulate the array you currently iterate over :thinking: 

You could pass `$language` by reference into the loop, that way you won't need this line anymore

```php
foreach ($languages as $languageId => &$language) {
```

What do you think?

silviokennecke replied:
I missed, that a child language must not have another child. Therefore, a recursive resolver is not necessary anyway. I just moved that to be handled by SQL.

Regarding your thought, what php does in that case: PHP uses a copy of `$languages[$languageId]` for `$language`, if not explicitly using the reference (as suggested).

silviokennecke committed:
replace resolver with sql

As child languages must not have other child languages, the recursive resolving turns out to be unnecessary.

aragon999 commented:
What is the difference to this PR? :)
https://github.com/shopware/platform/pull/2461

I mean I am totally in favor of this change and would like to see it happen, just curious. And maybe extend the discussion from the other PR, as I do not agree with the current outcome, as the `code` field is optional for inherited languages. 

Also it might be reasonable to add the tests from the other PR as well.

silviokennecke commented:
@aragon999 I totally missed that one, thanks for the hint. The issue from my point was (different from #2461) that the current code produces error 500 responses in case you use child languages - which was quite problematic at a customer's project when recently updating Shopware. To fix this also in the core, for me it would also be okay fixing it in the `LanguageLocaleProvider`, but it definitely makes more sense to fix it in the data source (so in the `LanguageLoader`), as suggested by @OliverSkroblin.

I'll have a more profound look into the other pr later this evening or tomorrow and check if I add the tests here as well.

aragon999 commented:
As I was a bit involved in the creation of the other PR, I can tell, that exactly the thing you mentioned was the reason for that PR.
We also started just by fixing the `getLocaleForLanguageId`, but than realized that the issue was also present in other cases (the `ids` function), it is just not stated like this in the PR message :)

The reasoning was than, that the mapping (`language` <-> `locale` via `translated_code_id`) should be unique, which makes no sense in my opinion, as you can only leave the translation code empty for inherited languages.

As I do not fully understand the reasoning, why the mapping should be unique in this case, it could also be that I misunderstand something, but the problem, which the other PR tried to solve, is exactly the same :) and I think the optimal way to fix this is via the `LanguageLoader`.

keulinho commented:
@aragon999 nice catch, i was not aware of the previous PR, that it does exactly the same thing, i only was aware that the previous PR started an internal discussion. But as this came already up the second time here i guess we have to fix something.

The reason the translation code id needs to be unique comes from the API/DAL as we allow writing translations by specifying the locale code so something like this works:
```php
$productRepository->upsert([
    [
        'id' => $id,
        'name' => [
            'en-GB' => 'test product',
            'de-DE' => 'Testprodukt',
       ],
   ],
], $context);
```
In that case i would say that the expected behaviour would be that the translation value will be written for the parent language and if you want to override a value for a child language (that inherits the translation code) you would have to provide the correct language id, either in the payload directly or set it in the context. As already described by @aragon999 in the previous PR.

@silviokennecke can you add a test case (in the `\Shopware\Core\Framework\Test\DataAbstractionLayer\Write\TranslationTest`) that verifies the behaviour i described regarding writing translated data? And please provide a second test case that you can write the translation over the child language over the id of that language.

keulinho commented:
@silviokennecke i just dug deeper into it and sadly the `LanguageLoader` is also used when we handle write payloads, that means in the current implementation you would get inconsistent writes because the DAL would write the translation values for the first language that it finds with that locale.

So i guess the best solution would be to switch back and solve the issue in the `LanguageLocaleCodeProvider` directly.

silviokennecke committed:
Revert "fix typo"

This reverts commit 30031b3fb974a8b718f318c3fe321bbf18fd9cd2.

silviokennecke committed:
Revert "revert locale code fix"

This reverts commit 33abecfe481b1c0137b2c19e2ee8ffdf533003c4.

silviokennecke committed:
Revert "replace resolver with sql"

This reverts commit 1726eb332d0354f173b24db696b181b5231b8f7a.

silviokennecke committed:
Revert "resolve locale code for child languages"

This reverts commit ceb75a431545d2cc80da61fc0a3082ec09ab8284.

keulinho commented:
@silviokennecke thanks again for the contribution! :blue_heart: Your changes are merged. :tada:
